File: PHP/src/TaxTable.php
<?php
namespace Gov;

// see https://github.com/brick/math
use Brick\Math\BigDecimal;

class TaxTable
{

    private $noTax;
    private $levels;

    public function __construct()
    {
        $this->noTax = BigDecimal::zero();
        $this->levels = [];
    }

    public function addProgression(BigDecimal $threshold, $percentage)
    {
        $decimalPercentage = BigDecimal::of($percentage)
            ->exactlyDividedBy(BigDecimal::of(100));
        array_push($this->levels, new IncomeRange($threshold, $decimalPercentage));
    }

    public function taxFor(BigDecimal $income)
    {
        foreach ($this->levels as $level) {
            if ($income->compareTo($level->threshold) >= 0) {
                return $income->multipliedBy($level->percentage);
            }
        }
        return $this->noTax;
    }

}

File: Python/tax_table.py
from income_range import IncomeRange


class TaxTable:

    def __init__(self):
        self._no_tax = 0
        self._levels = []

    def add_progression(self, threshold, percentage):
        decimal_percentage = percentage / 100
        self._levels.append(IncomeRange(threshold, decimal_percentage))

    def taxFor(self, income):
        for level in self._levels:
            if income >= level.threshold:
                return income * level.percentage
        return self._no_tax
